src/GoTrueClient.ts

```diff
@@ -2442,6 +2474,150 @@ export default class GoTrueClient {
+      const { data: sessionData, error: sessionError } = sessionResult
+      if (sessionError) {
+        return { data: null, error: sessionError }
+      }
+      // Single Step enroll
+      const webAuthn = this._getWebAuthnRpDetails()
+
+      const { data: challengeData, error: challengeError } = await this._challenge({
+        factorId: webAuthnFactor.id,
+        webAuthn,
+      })
+      if (
+        !challengeData ||
+        !(challengeData.type === 'webauthn' && challengeData?.credential_request_options)
+      ) {
+        return {
+          data: null,
+          error: new Error('Invalid challenge data for WebAuthn'),
+        }
+      }
+      const challengeOptions = challengeData?.credential_request_options.publicKey
+      const finalCredential = await startAuthentication(challengeOptions)
+      const verifyWebAuthnParams = { ...webAuthn, assertionResponse: finalCredential }
+
+      const { data, error } = await _request(
+        this.fetch,
+        'POST',
+        `${this.url}/factors/${webAuthnFactor.id}/verify`,
+        {
+          body: {
+            challenge_id: challengeData.id,
+            web_authn: {
+              rp_id: verifyWebAuthnParams.rpId,
+              rp_origins: verifyWebAuthnParams.rpOrigins,
+              assertion_response: verifyWebAuthnParams.assertionResponse,
+            },
+          },
+          headers: this.headers,
+          jwt: sessionData?.session?.access_token,
+        }
+      )
+      if (error) {
+        return { data: null, error }
+      }
+
+      await this._saveSession({
+        expires_at: Math.round(Date.now() / 1000) + data.expires_in,
+        ...data,
+      })
+      await this._notifyAllSubscribers('MFA_CHALLENGE_VERIFIED', data)
+      return { data, error }
+    })
+  }
+
+  private async _verifyWebAuthnCreation(
+    params: MFAVerifySingleStepWebAuthnParams
+  ): Promise<AuthMFAVerifyResponse> {
+    return this._useSession(async (sessionResult) => {
+      const { data: sessionData, error: sessionError } = sessionResult
+      if (sessionError) return { data: null, error: sessionError }
+
+      if (!params.webAuthn) {
+        return { data: null, error: new AuthError('Invalid MFA parameters') }
+      }
+
+      const { data, error } = await _request(
+        this.fetch,
+        'POST',
+        `${this.url}/factors/${params.factorId}/verify`,
+        {
+          body: {
+            challenge_id: params.challengeId,
+            web_authn: {
+              rp_id: params.webAuthn.rpId,
+              rp_origins: params.webAuthn.rpOrigins,
+              creation_response: params.webAuthn.creationResponse,
+            },
+          },
+          headers: this.headers,
+          jwt: sessionData?.session?.access_token,
+        }
+      )
+
+      if (error) return { data: null, error }
+
+      await this._saveSession({
+        expires_at: Math.round(Date.now() / 1000) + data.expires_in,
```

Comment:
Let's move this in `_saveSession`.

Reply:
I think we have a slightly different computation in the `_saveSession` calls elsewhere. Will try to re-concile though.

Reply:
Yeah a bit harder to compute in `_saveSession` as there are places where we pass in the session with the `expires_at` field unedited [like here](https://github.com/supabase/auth-js/blob/master/src/GoTrueClient.ts#L1257) so we can't do the adjustment without affecting existing calls.

Let me know if there's a workaround missed though.